### PR TITLE
deprecate grib

### PR DESCRIPTION
**Pull request by marqh**
deprecation warning for use of not_strict GRIB loading

### CODE REVIEW

File: lib/iris/fileformats/grib/__init__.py [hunk 1]
```diff
@@ -924,6 +924,15 @@ def load_cubes(filenames, callback=None, regularise=True, auto_regularise=None):
             {'regularise': regularise},
             iris.fileformats.grib._load_convert.convert, None)
     else:
+        msg = ('This GRIB loader is deprecated and will be removed in '
+              'a future release.  Please consider using the new '
+              'GRIB loader, by setting the :class:`iris.Future` '
+              'option `strict_grib_load` to True; e.g.:\n.'
```
Comment (esc24) [inline lib/iris/fileformats/grib/__init__.py]: Extra `.` after the `\n`

**Comment by ajdawson** [comment]
Can we hold off on merging this until a bit of discussion has occurred on the thread Mark sent round to the groups?

**Comment by marqh** [comment]
Hi AJ

these discussions will take place over some time.  I would very much like the deprecation warning on master to encourage engagement in these discussions

I don't see what will come out of the discussions in short time that affects the warning

the guidelines on deprecation talk about 6 months and 2 release so we have a long time to consider implications and code changes

this is getting things started really
mark

**Comment by esc24** [comment]
@ajdawson - I'm keen to get this warning out as soon as possible, certainly before the next release, so people are aware that this change will be coming at some point. At the moment, I suspect users are not aware of it so aren't opting in.

**Comment by ajdawson** [comment]
OK, fair enough. Instead would you clarify for me if the removal of this loader means the removal of GRIB1 support from Iris?

**Comment by marqh** [comment]
> OK, fair enough. Instead would you clarify for me if the removal of this loader means the removal of GRIB1 support from Iris?

as far as I am concerned absolutely not

i expect the new loader to have GRIB1 support; I would like this to be in place before it becomes the default choice

**Comment by ajdawson** [comment]
But the new loader now does not support GRIB1?

**Comment by marqh** [comment]
> But the new loader now does not support GRIB1?

At present it only supports GRIB2, not GRIB1.  It needs to support GRIB1 loading but the code has not yet been written

**Comment by esc24** [comment]
> But the new loader now does not support GRIB1?

I didn't appreciate this.

**Comment by ajdawson** [comment]
I'm just a bit concerned about deprecating something that has no replacement. A user loading GRIB1 will see a warning suggesting they try the new loader, which won't work for their files, and also tell them the old loader (the only thing that works on their files) will be going away. I'm not sure it sends the right message...

**Comment by marqh** [comment]
we need to know which GRIB1 messages to concentrate efforts on.  Currently we don't support all of GRIB1 but which areas work and which don't is buried in the code.

I want explicit knowledge of what is supported and for that we need user feedback.  A deprecation warning is, in my view, the best way to encourage engagement

In your scenario I would hope user would post on the group saying 'didn't work for me, here are my files': then we have a chance of focusing effort to support the important cases

**Comment by esc24** [comment]
Can we push grib2 down the new pipeline and grib1 down the old?

**Comment by marqh** [comment]
> Can we push grib2 down the new pipeline and grib1 down the old?

We can, this is a very valid back up plan.  I think we should keep this as an option but not implement it yet.  
I would very much like to see what feedback the deprecation warning throws up.
If this, and time constraints, lead us to choose this path then that would be just fine.  
It seems a lost opportunity to do this now and not get feedback on the requirements for GRIB1 support

**Comment by rhattersley** [comment]
@ajdawson - thanks for mentioning the GRIB1 issue. :+1: I knocked up this deprecation plan with @marqh a few days ago, but the GRIB1 problem didn't occur to me - sorry! I agree that issuing deprecation warnings for functionality that has no alternative is sending the wrong message. So I'm :-1: on this change as it stands. @esc24's suggestion of only deprecating for GRIB2 seems sensible.

**Comment by marqh** [comment]
OK, I have updated the warning so that it only triggers if a GRIB edition 2 message is loaded

it triggers on every message but warning suppression should limit the user view to 1

**Comment by marqh** [comment]
please may I have follow up comments on this?

have I addressed the concerns such that this can be merged?

**Comment by esc24** [comment]
This looks better to me. I'm keen to see a new release in the very near future and I'd be happy for this GRIB2-only deprecation of the old code path to go in. @ajdawson - what do you think about this modified PR?

**Comment by ajdawson** [comment]
I'm OK with this.

**Comment by esc24** [comment]
@marqh - would you mind rebasing and squashing so this is a single commit that only modifies one file?

**Comment by marqh** [comment]
rebased

File: lib/iris/fileformats/grib/__init__.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-# (C) British Crown Copyright 2010 - 2014, Met Office
+# (C) British Crown Copyright 2010 - 2015, Met Office
```
Comment (ajdawson) [inline lib/iris/fileformats/grib/__init__.py]: This shouldn't be here.

Comment (esc24) [inline lib/iris/fileformats/grib/__init__.py]: Agreed - as mentioned above, this PR should modify a single file.

**Comment by marqh** [comment]
hopefully this is clean and good to merge now

**Comment by rhattersley** [comment]
Looks like everyone is OK with this now ... ?

**Comment by ajdawson** [comment]
:+1:

**Comment by rhattersley** [comment]
> :+1:

Thanks @ajdawson.

@esc24 has moved on to greener pastures so unsurprisingly he doesn't seem to be following as closely any more. Since his comments above suggest he was now in favour, rather than wait I'm going to merge now and he can always complain afterwards. :wink:

**Pull request merged**